Linking against `-lz` seems missing

Recently [an error has been reported](https://bugs.debian.org/1086362) on Debian builds:
```
[100%] Linking CXX executable MMB
/usr/bin/cmake -E cmake_link_script CMakeFiles/MMB_EXEC.dir/link.txt --verbose=1
/usr/bin/c++ -D BuildNtC -D USE_OPENMM -g -O2 -ffile-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -O0 -fvisibility=hidden -Wall -pedantic -O3 -DNDEBUG -Wl,-z,relro CMakeFiles/MMB_EXEC.dir/src/RNABuilder.cpp.o -o MMB   -L/<<PKGBUILDDIR>>/SYSTEM  -Wl,-rpath,/<<PKGBUILDDIR>>/SYSTEM:/<<PKGBUILDDIR>>/obj-x86_64-linux-gnu: libMMB.so.4.0 /usr/lib/x86_64-linux-gnu/libSimTKmolmodel.so.3.1 /usr/lib/x86_64-linux-gnu/libSimTKsimbody.so.3.7 /usr/lib/x86_64-linux-gnu/libSimTKmath.so.3.7 /usr/lib/x86_64-linux-gnu/libSimTKcommon.so.3.7 /usr/lib/x86_64-linux-gnu/libblas.so /usr/lib/x86_64-linux-gnu/liblapack.so /usr/lib/x86_64-linux-gnu/libblas.so /usr/lib/x86_64-linux-gnu/liblapack.so -lpthread -lrt -ldl -lm
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzerror'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzgetc'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzgets'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzopen'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzread'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzungetc'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzclose_r'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzbuffer'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzeof'
collect2: error: ld returned 1 exit status
```
To me it seems that adding `-lz` to `LDFLAGS` should fix the issue, although I have not tried yet.

Hi Andrius,

I usually compile using cmake. wouldn't it be easier to follow that ? I
don't think I've gotten this error that way.

I wonder if these libraries are required by MMB. I wonder if maybe it is
only that gemmi is not being installed properly? gemmi has gz.hpp which is
the only place that any gz operations are being done.

Sam

On Sat, Nov 9, 2024 at 8:28 AM Andrius Merkys ***@***.***>
wrote:

> Recently an error has been reported <https://bugs.debian.org/1086362> on
> Debian builds:
>
> [100%] Linking CXX executable MMB
> /usr/bin/cmake -E cmake_link_script CMakeFiles/MMB_EXEC.dir/link.txt --verbose=1
> /usr/bin/c++ -D BuildNtC -D USE_OPENMM -g -O2 -ffile-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -O0 -fvisibility=hidden -Wall -pedantic -O3 -DNDEBUG -Wl,-z,relro CMakeFiles/MMB_EXEC.dir/src/RNABuilder.cpp.o -o MMB   -L/<<PKGBUILDDIR>>/SYSTEM  -Wl,-rpath,/<<PKGBUILDDIR>>/SYSTEM:/<<PKGBUILDDIR>>/obj-x86_64-linux-gnu: libMMB.so.4.0 /usr/lib/x86_64-linux-gnu/libSimTKmolmodel.so.3.1 /usr/lib/x86_64-linux-gnu/libSimTKsimbody.so.3.7 /usr/lib/x86_64-linux-gnu/libSimTKmath.so.3.7 /usr/lib/x86_64-linux-gnu/libSimTKcommon.so.3.7 /usr/lib/x86_64-linux-gnu/libblas.so /usr/lib/x86_64-linux-gnu/liblapack.so /usr/lib/x86_64-linux-gnu/libblas.so /usr/lib/x86_64-linux-gnu/liblapack.so -lpthread -lrt -ldl -lm
> /usr/bin/ld: libMMB.so.4.0: undefined reference to `gzerror'
> /usr/bin/ld: libMMB.so.4.0: undefined reference to `gzgetc'
> /usr/bin/ld: libMMB.so.4.0: undefined reference to `gzgets'
> /usr/bin/ld: libMMB.so.4.0: undefined reference to `gzopen'
> /usr/bin/ld: libMMB.so.4.0: undefined reference to `gzread'
> /usr/bin/ld: libMMB.so.4.0: undefined reference to `gzungetc'
> /usr/bin/ld: libMMB.so.4.0: undefined reference to `gzclose_r'
> /usr/bin/ld: libMMB.so.4.0: undefined reference to `gzbuffer'
> /usr/bin/ld: libMMB.so.4.0: undefined reference to `gzeof'
> collect2: error: ld returned 1 exit status
>
> To me it seems that adding -lz to LDFLAGS should fix the issue, although
> I have not tried yet.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/samuelflores/MMB/issues/26>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABQ44PMHZ3TBERYU46A6HFDZ7W2RVAVCNFSM6AAAAABRO3MU2SVHI2DSMVQWIX3LMV43ASLTON2WKOZSGY2DKNZVGEZDEMY>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

Now the problem seems to be occurring down at the molmodel level. I get the following on the "samuelflores/mmb-ubuntu" docker image, after updating to. latest gemmi and molmodel. Maybe you are building molmodel to a specific release?

root@0f646d5f78ff:/github/molmodel/build# make    
[ 25%] Built target SimTKmolmodel
[ 26%] Linking CXX executable TestAtomTargetLocations
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzerror'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzgetc'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzgets'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzopen'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzread'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzungetc'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzclose_r'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzbuffer'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzeof'
collect2: error: ld returned 1 exit status
make[2]: *** [tests/CMakeFiles/TestAtomTargetLocations.dir/build.make:105: tests/TestAtomTargetLocations] Error 1
make[1]: *** [CMakeFiles/Makefile2:1030: tests/CMakeFiles/TestAtomTargetLocations.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
root@0f646d5f78ff:/github/molmodel/build#

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -312,6 +312,7 @@ if (BUILD_SHARED)
     target_link_libraries(
         ${MMB_SHARED_TARGET}
         PRIVATE ${LIBMMB_SHARED_TARGET}
+        PRIVATE ${ZLIB_LIBRARIES}
     )
 endif ()
 

